src/db_impl.cc

```diff
@@ -762,6 +763,80 @@ void TitanDBImpl::ReleaseSnapshot(const Snapshot* snapshot) {
   db_->ReleaseSnapshot(snapshot);
 }
 
+Status TitanDBImpl::DisableFileDeletions() {
+  // Disable base DB file deletions.
+  Status s = db_impl_->DisableFileDeletions();
+  if (!s.ok()) {
+    return s;
+  }
+
+  int count = 0;
+  {
+    // Hold delete_titandb_file_mutex_ to make sure no
+    // PurgeObsoleteFiles job is running.
+    MutexLock l(&delete_titandb_file_mutex_);
+    count = ++disable_titandb_file_deletions_;
+  }
+
+  ROCKS_LOG_INFO(db_options_.info_log,
+                 "Disalbed blob file deletions. count: %d", count);
+  return Status::OK();
+}
+
+Status TitanDBImpl::EnableFileDeletions(bool force) {
+  // Enable base DB file deletions.
+  Status s = db_impl_->EnableFileDeletions(force);
+  if (!s.ok()) {
+    return s;
+  }
+
+  int count = 0;
+  {
+    MutexLock l(&delete_titandb_file_mutex_);
+    if (force) {
+      disable_titandb_file_deletions_ = 0;
+    } else if (disable_titandb_file_deletions_ > 0) {
+      count = --disable_titandb_file_deletions_;
+    }
+    assert(count >= 0);
+  }
+
+  ROCKS_LOG_INFO(db_options_.info_log, "Enabled blob file deletions. count: %d",
+                 count);
+  return Status::OK();
+}
+
+Status TitanDBImpl::GetTitanLiveFiles(std::vector<std::string>& base_ret,
+                    uint64_t* base_manifest_file_size,
+                    std::vector<std::string>& titan_ret,
+                    uint64_t* titan_manifest_file_size,
+                    bool flush_memtable) {
+  // We can not hold titan db mutex in the beginning to avoid updates to base DB 
+  // during call GetLiveFiles, because in DB::GetLiveFiles maybe happen flush 
+  // memtable. Flush memtable operation hold titan db mutex, this can cause dead lock.
+  // So: 
+  // 1. DisableFileDeletions, this will prevent blob files being deleted.
+  // 2. Call db_->GetLiveFiles without mutex to get all live file about base db.
+  // 3. Collect all file（live + obsolete）about titandb
+  // 4. EnableFileDeletions
+  // This will include redundant blob files, but hopefully not a lot of them, and on 
+  // restart Titan will recalculate GC stats and GC out those redundant blob files.
+  Status s = DisableFileDeletions();
+  if (!s.ok()) {
+    return s;
+  }
+  
+  s = db_->GetLiveFiles(base_ret, base_manifest_file_size, flush_memtable);
+  if (!s.ok()) {
+    return s;
+  }
+
+  MutexLock l(&mutex_);
+  blob_file_set_->GetAllFiles(&titan_ret, titan_manifest_file_size);
```

Comment:
After a second thought, Disbaling file deletion doesn't solve the problem. It's just keep the blob file from being pyhsically deleted, but blob files are logically deleted. Considing a Titan GC is happened between db_->GetLiveFiles and blob_file_set_->GetAllFiles, the titan_manifest_file_size and base_manifest_file_size are not fetched in atomic, so the blob file is logically deleted whereas rocksdb still refer to it. When reopen the checkpoint, titan will deleted the logical deleted file at start. Then it will meet a blob file not found issue.